engine/src/test/java/org/hibernate/search/test/projection/ProjectionConversionTest.java

```diff
@@ -0,0 +1,241 @@
+ * when the projection handling code is not careful about this.
+ *
+ * @author Sanne Grinovero <sanne@hibernate.org> (C) 2015 Red Hat Inc.
+ */
+@TestForIssue(jiraKey = "HSEARCH-1786")
+public class ProjectionConversionTest {
+
+	@Rule
+	public SearchFactoryHolder sfHolder = new SearchFactoryHolder( ExampleEntity.class );
+
+	@Before
+	public void storeTestData() {
+		ExtendedSearchIntegrator searchFactory = sfHolder.getSearchFactory();
+		ExampleEntity entity = new ExampleEntity();
+		entity.id = 1l;
+		entity.someInteger = 5;
+		entity.longEncodedAsText = 20l;
+		entity.customBridgedKeyword = "lowercase-keyword";
+
+		ExampleEntity embedded = new ExampleEntity();
+		embedded.id = 2l;
+		embedded.someInteger = 6;
+		embedded.longEncodedAsText = 21l;
+		embedded.customBridgedKeyword = "another-lowercase-keyword";
+
+		ConflictingMappedType second = new ConflictingMappedType();
+		second.id = "a string";
+		second.customBridgedKeyword = 17l;
+
+		entity.embedded = embedded;
+		entity.second = second;
+
+		Work work = new Work( entity, entity.id, WorkType.ADD, false );
+		TransactionContextForTest tc = new TransactionContextForTest();
+		searchFactory.getWorker().performWork( work, tc );
+		tc.end();
+	}
+
+	@Test
+	public void projectingExplicitId() {
+		projectionTestHelper( ProjectionConstants.ID, Long.valueOf( 1l ) );
+	}
+
+	@Test
+	public void projectingIdByPropertyName() {
+		projectionTestHelper( "id", Long.valueOf( 1l ) );
+	}
+
+	@Test
+	public void projectingIdOnOverloadedMapping() {
+		projectionTestHelper( "stringTypedId", Long.valueOf( 1l ) );
+	}
+
+	@Test
+	public void projectingIntegerField() {
+		projectionTestHelper( "someInteger", Integer.valueOf( 5 ) );
+	}
+
+	@Test
+	public void projectingUnknownField() {
+		projectionTestHelper( "someNonExistingField", null );
+	}
+
+	@Test
+	public void projectionWithCustomBridge() {
+		projectionTestHelper( "customBridgedKeyword", "lowercase-keyword" );
+	}
+
+	@Test
+	public void projectingEmbeddedIdByPropertyName() {
+		projectionTestHelper( "embedded.id", Long.valueOf( 2l ) ); //FAILS OFTEN
+	}
+
+	@Test
+	public void projectingEmbeddedIdOnOverloadedMapping() {
+		projectionTestHelper( "embedded.stringTypedId", Long.valueOf( 2l ) );
+	}
+
+	@Test
+	public void projectingEmbeddedWithCustomBridge() {
+		projectionTestHelper( "embedded.customBridgedKeyword", "another-lowercase-keyword" );
+	}
+
+	@Test
+	public void projectingNotIncludedEmbeddedField() {
+		projectionTestHelper( "embedded.someInteger", null );
+	}
+
+	@Test
+	public void projectingOnConflictingMappingEmbeddedField() {
+		projectionTestHelper( "second.customBridgedKeyword", Long.valueOf( 17l ) );
+	}
+
+	@Test
+	public void projectingOnConflictingMappedIdField() {
+		projectionTestHelper( "second.id", "a string" );
+	}
+
+	@Test
+	public void concurrentMixedProjections() throws Exception {
+		//The point of this test is to "simultaneously" project multiple different types
+		new ConcurrentRunner( 1000, 20,
```

Comment:
It turns out that concurrency wasn't the issue, but it's good to test for this down the road.
Since it would only fail "occasionally", I suspected we could have an issue with mixing up parallel projection requests

It only takes 0.2 seconds to run (especially after fixing the logger threshold) so I'd rather keep it.